include/cetl/variable_length_array.hpp

```diff
@@ -0,0 +1,2065 @@
+/// First, the `max_size_max` maximum bound allows the array to enforce a maximum size whether or not the allocator
+/// properly implements the std::allocator_traits::max_size protocol (std::pmr::polymorphic_allocator, for example).
+/// This allows use of an allocator that is backed by statically allocated memory:
+/// @snippet{trimleft} example_05_variable_length_array_vs_vector.cpp example_exact_fit
+///
+/// Even if cetl::VariableLengthArray::reserve() is not used the implementation will still use less memory then its
+/// STL counterpart:
+/// @snippet{trimleft} example_05_variable_length_array_vs_vector.cpp example_tight_fit_0
+/// @snippet{trimleft} example_05_variable_length_array_vs_vector.cpp example_tight_fit_1
+///
+/// Finally, when exceptions are disabled the cetl::VariableLengthArray will not exhibit undefined behaviour like
+/// std::vector does:
+/// @snippet{trimleft} example_05_variable_length_array_vs_vector.cpp example_no_exceptions
+///
+/// @tparam  T           The type of elements in the array.
+/// @tparam Allocator    The allocator type to use for all allocations.
+///
+template <typename T, typename Allocator>
+class VariableLengthArray : protected VariableLengthArrayBase<T, Allocator>
+{
+protected:
+    using Base = VariableLengthArrayBase<T, Allocator>;
+    using Base::data_;
+    using Base::size_;
+    using Base::capacity_;
+    using Base::max_size_max_;
+    using Base::alloc_;
+
+public:
+    ///
+    /// STL-like declaration of the container's storage type.
+    ///
+    using value_type = typename Base::value_type;
+
+    ///
+    /// STL-like declaration of pointer type.
+    ///
+    using pointer = typename std::add_pointer_t<value_type>;
+
+    ///
+    /// STL-like declaration of the iterator type.
+    ///
+    using iterator = typename std::add_pointer_t<value_type>;
+
+    ///
+    /// STL-like declaration of the const iterator type.
+    ///
+    using const_iterator = typename std::add_pointer_t<typename std::add_const_t<value_type>>;
+
+    ///
+    /// STL-like declaration of the container's difference type.
+    ///
+    using difference_type = typename Base::difference_type;
+
+    ///
+    /// STL-like declaration of the container's size type.
+    ///
+    using size_type = typename Base::size_type;
+
+    ///
+    /// Per the std::uses_allocator protocol.
+    ///
+    using allocator_type = typename Base::allocator_type;
+
+    ///
+    /// STL-like declaration of reference type.
+    ///
+    using reference = typename std::add_lvalue_reference_t<value_type>;
+
+    ///
+    /// STL-like declaration of constant-reference type.
+    ///
+    using const_reference = typename std::add_const_t<std::add_lvalue_reference_t<value_type>>;
+
+    /// Required constructor for std::uses_allocator protocol.
+    /// @param alloc Allocator to use for all allocations.
+    explicit VariableLengthArray(const allocator_type& alloc) noexcept
```

Comment:
Is there a reason we don't default Allloc=std::allocator<T> and alloc=Alloc() here? (like std::vector does)

Reply:
Yeah, CETL deviates from the standard where it requires system resources. heap memory provided by the system and requested via SBRK are system resources. This is why we isolate the `new_delete_resource` in the `sys/memory_resource.hpp` header.